libs/switches.liq

```diff
@@ -0,0 +1,166 @@
+# At the beginning of each track, select the first ready child.
+# @category Source / Track Processing
+# @param ~id Force the value of the source ID.
+# @param ~override Metadata field which, if present and containing a float, overrides the `transition_length` parameter.
+# @param ~replay_metadata Replay the last metadata of a child when switching to it in the middle of a track.
+# @param ~track_sensitive Re-select only on end of tracks.
+# @param ~transition_length Maximun transition duration.
+# @param ~transitions Transition functions, padded with `fun (x,y) -> y` functions.
+def fallback(~id="", ~override="liq_transition_length", ~replay_metadata=true,
+           ~track_sensitive=true, ~transition_length=5., ~transitions=[],
+           ~weights=[], sources) =
+  def add_condition(s) =
+    ({true}, s)
+  end
+
+  switch(id=id, override=override, replay_metadata=replay_metadata, transitions=transitions,
+         track_sensitive=track_sensitive, transition_length=transition_length,
+         list.map(add_condition, sources))
+end
+
+# Rotate between the sources.
+# @category Source / Track Processing
+# @param ~id Force the value of the source ID.
+# @param ~override Metadata field which, if present and containing a float, overrides the `transition_length` parameter.
+# @param ~replay_metadata Replay the last metadata of a child when switching to it in the middle of a track.
+# @param ~transition_length Maximun transition duration.
+# @param ~transitions Transition functions, padded with `fun (x,y) -> y` functions.
+# @param ~weights Weights of the children (padded with 1), defining for each child how many tracks are played from it per round, if that many are actually available.
+def rotate(~id="", ~override="liq_transition_length", ~replay_metadata=true,
+           ~transition_length=5., ~transitions=[],
+           ~weights=[], sources) =
+  weights = list.map(to_int_getter, weights)
+  picked_index = ref -1
+  tracks_played = ref 0
+  default_weight = fun () -> 1
+  fail = fallback([])
+
+  def pick() =
+    tracks_played := 0
+
+    if list.exists(source.is_ready, sources) then
+      def rec f(index) =
+        s = list.nth(default=fail, sources, index)
+        if source.is_ready(s) then
+          picked_index := index
+        else
+          f((index + 1) mod list.length(sources))
```

Comment:
Our streaming model call `is_ready` before calling `get_frame` and expects `get_frame` to succeed when `is_ready` previously returned `true`. Although probably not enforced enough, there is an implicit assumption here that the next call to `get_frame` should always succeed when a source returns `is_ready = true`. This would apply here since those predicates are all computed before any call to `get_frame` is executed.

Reply:
```
       (** [is_ready] tells you if [get] can be called. *)
       method virtual is_ready : bool
```